Add `bazel run` support to `macos_application`

rahul-malik opened the pull request:
Addresses https://github.com/bazelbuild/rules_apple/issues/357

Primary changes to enable this:
- Add a new runner script for executing macos binaries (update
rule_factory to make this template available)
- Add method to run_support.bzl to enable execution of applications
without a simulator
- Update macos_application implementation to pass the archive as
runfiles, generate an executable script from the new template

RELNOTES: Adds support for executing `macos_application` targets using `bazel run`

rahul-malik commented on apple/internal/templates/macos.template.sh:
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+# Copyright 2015 The Bazel Authors. All rights reserved.
```
I copied this header, let me know if I need to strip it, update the copyright year, etc

rahul-malik commented on apple/internal/rule_factory.bzl:
```diff
@@ -137,6 +137,11 @@ _COMMON_PRIVATE_TOOL_ATTRS = dicts.add(
+            cfg = "host",
+            allow_single_file = True,
+            default = Label("@build_bazel_rules_apple//apple/internal/templates:macos_sim_template"),
+        ),
```
The above `_runner_template` mentions supporting multiple platforms. I'm not sure what the exact solution they were imagining was so I just added another entry here.

keith commented on apple/internal/rule_factory.bzl:
```diff
@@ -137,6 +137,11 @@ _COMMON_PRIVATE_TOOL_ATTRS = dicts.add(
+        "_macos_runner_template": attr.label(
+            cfg = "host",
+            allow_single_file = True,
+            default = Label("@build_bazel_rules_apple//apple/internal/templates:macos_sim_template"),
```
Sim in the file name probably doesn't make sense here

rahul-malik replied:
Thanks! Will change

rahul-malik replied:
Removed, thanks!

keith commented on apple/internal/run_support.bzl:
```diff
@@ -37,7 +37,25 @@ def _register_simulator_executable(ctx, output):
+        template = ctx.file._macos_runner_template,
+        substitutions = {
+            "%app_name%": ctx.label.name,
+            "%ipa_file%": outputs.archive(ctx).short_path,
```
Ipa file probably isn't the right term

rahul-malik replied:
Yeah I’m not sure what to call this. “app_bundle_path”?

keith replied:
That or app_path sounds fine to me

rahul-malik replied:
Updated!

rahul-malik commented on apple/internal/templates/BUILD:
```diff
@@ -18,6 +18,15 @@ filegroup(
 )
 
+filegroup(
+    name = "macos_sim_template",
```
Will update to remove “sim”

rahul-malik commented on apple/internal/templates/macos.template.sh:
```diff
@@ -0,0 +1,45 @@
+readonly APP_DIR="${TEMP_DIR}/extracted_app"
+mkdir "${APP_DIR}"
+
+if [ -d '%ipa_file%' ]; then
```
Will update references to ipa_file with a better name

sergiocampama commented on apple/internal/templates/macos.template.sh:
```diff
@@ -0,0 +1,45 @@
+
+if [ -d '%app_path%' ]; then
+  # App bundles are directories with the .app extension
+  mv '%app_path%' "${APP_DIR}"
```
is it needed to move the app into a different location? how about:

```
if -d app_path
   APP_DIR=app_path
else:
   tempdir = mktemp bazel_temp.xxx
   unzip -qq app_path -d temp_dir
   APP_DIR=tempdir/app_name.app

bundle_info_plist = app_dir/Contents/Info.plist
bundle_executable=...

app_dir/contents/MacOS/bundle_executable
```

this way, a single directory is created (besides the unzip) and only if the app was zipped

rahul-malik replied:
Sounds good. Nope this was just a quick thing to do when I was adding support for tree artifacts but I agree it’s unnecessary

rahul-malik replied:
@sergiocampama do you think that checking the extension of `app_path` is required here or is checking that it's a directory sufficient for now?

rahul-malik commented on apple/internal/run_support.bzl:
```diff
@@ -37,7 +37,25 @@ def _register_simulator_executable(ctx, output):
         },
     )
 
+def _register_app_executable(ctx, output):
```
A nit about your nit :)

Should it be _register_macos_executable since the result of this function is writing the executable file and it’s consistent with the sim executable method in the same file?

sergiocampama replied:
sounds good 👍

rahul-malik commented on apple/internal/run_support.bzl:
```diff
@@ -37,7 +37,25 @@ def _register_simulator_executable(ctx, output):
     )
 
+def _register_app_executable(ctx, output):
+    """Registers an action that runs the bundled app.
```
Will update!

sergiocampama commented:
one final thing, if in the first comment you use RELNOTES: , that line will appear in the release notes for the next version :)

rahul-malik commented:
@sergiocampama @thomasvl - I just rebased master, addressed comments and added `RELNOTES` to my commit and in the first comment of this PR. Let me know if you have any more feedback I need to address

rahul-malik commented:
@sergiocampama
> one final nit
I don't see a comment to address? Was this just in the text field when you approved the PR?

sergiocampama commented on apple/internal/run_support.bzl:
```diff
@@ -37,7 +37,25 @@ def _register_simulator_executable(ctx, output):
+        is_executable = True,
+        template = ctx.file._macos_runner_template,
+        substitutions = {
+            "%app_name%": ctx.label.name,
```
sorry, it must have not saved... you'll need to use `bundling.bundle_name(ctx)`  here in order to get the bundle name as as configured in the `bundle_name` attribute.

rahul-malik replied:
@sergiocampama - No worries! Just updated, PTALe
